tests/test_core.py

```diff
@@ -627,10 +627,54 @@ def test_interpolate(test_pd_df):
     assert all([True if isstr(i) else ~np.isnan(i) for i in df.data.foo])
 
 
+def test_interpolate_time_exists(test_pd_df):
+    _df = test_pd_df.copy()
+    df = IamDataFrame(_df)
+    df.interpolate(2005)
+    obs = df.filter(year=2005).data['value'].reset_index(drop=True)
+    exp = pd.Series([1.0, 0.5, 2.0], name='value')
+    pd.testing.assert_series_equal(obs, exp)
+
+
+def test_interpolate_multiple_times(test_pd_df):
+    _df = test_pd_df.copy()
+    df = IamDataFrame(_df)
+    df.interpolate([2007, 2008])
+    obs = df.filter(year=[2007, 2008]).data['value'].reset_index(drop=True)
+    exp = pd.Series([3, 4, 1.5, 2, 4, 5], name='value')
+    pd.testing.assert_series_equal(obs, exp)
+
+
+def test_interpolate_full_example():
+    cols = ['model_a', 'scen_a', 'World']
+    df = IamDataFrame(pd.DataFrame([
+        cols + ['all', 'EJ/yr', 0, 1, 6., 10],
+        cols + ['last', 'EJ/yr', 0, 0.5, 3, np.nan],
+        cols + ['first', 'EJ/yr', 0, np.nan, 2, 7],
+        cols + ['middle', 'EJ/yr', 0, 1, np.nan, 7],
+        cols + ['first two', 'EJ/yr', 0, np.nan, np.nan, 7],
+        cols + ['last two', 'EJ/yr', 0, 1, np.nan, np.nan],
+    ], columns=IAMC_IDX + [2000, 2005, 2010, 2017],
+    ))
+    exp = IamDataFrame(pd.DataFrame([
+        cols + ['all', 'EJ/yr', 0, 1, 6., 7.142857, 10],
+        cols + ['last', 'EJ/yr', 0, 0.5, 3, np.nan, np.nan],
+        cols + ['first', 'EJ/yr', 0, 1., 2, 3.428571, 7],
+        cols + ['middle', 'EJ/yr', 0, 1, np.nan, 4.5, 7],
+        cols + ['first two', 'EJ/yr', 0, 2.058824, np.nan, 4.941176, 7],
+        cols + ['last two', 'EJ/yr', 0, 1, np.nan, np.nan, np.nan],
+    ], columns=IAMC_IDX + [2000, 2005, 2010, 2012, 2017],
+    ))
+    obs = df.interpolate([2005, 2012], inplace=False)
+    assert_iamframe_equal(obs, exp)
+
+
 def test_interpolate_extra_cols():
     # check hat interpolation with non-matching extra_cols has no effect (#351)
     EXTRA_COL_DF = pd.DataFrame([
         ['foo', 2005, 1],
+        ['foo', 2010, 2],
```

Comment:
@Rlamboll and @danielhuppmann I had to update this test to get correct behavior. The current implementation uses the wide-form `timeseries` object, which treats each index  value (row) as an indepedent observation. Therefore, the `foo` and `bar` entries here are each interpolated. If I left it as is, I would get an error, because there are too few datapoints to interpolate (i.e., 1 each)

Reply:
OK, I believe this is consistent with the current state as of the discussion in #351. It's unclear whether this is the desired behaviour or not, but it's what we're used to.

Reply:
Slightly concerned that this behavior might throw users that have an IamDataFrame where just one timeseries-row has only one datapoint...

Reply:
For the moment, an error will be raised if trying to interpolate with a single data point. I suggest we allow this, as I am not sure we 'want' to support a use case of pyam data with a single data point (this is what metadata is for, right)? @danielhuppmann I will let you mark if resolved.

Reply:
I'm worried if a user has a table like the one below and wants to interpolate for 2015.

| variable       | 2010 | 2020 | 2030 |
|----------------|------|------|------|
| Primary Energy | 1    | 2    | 3    |
| Population     | 5    | nan  | nan  |

I would expect a return-object like

| variable       | 2010 | 2015 | 2020 | 2030 |
|----------------|------|------|------|------|
| Primary Energy | 1    | 1.5 | 2    | 3    |
| Population     | 5    | nan | nan  | nan  |